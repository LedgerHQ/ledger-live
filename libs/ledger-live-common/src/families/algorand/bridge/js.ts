import { makeAccountBridgeReceive } from "../../../bridge/jsHelpers";
import type { AlgorandTransaction } from "../types";
import { sync, scanAccounts } from "../js-synchronization";
import {
  createTransaction,
  prepareTransaction,
} from "../js-prepareTransaction";
import { estimateMaxSpendable } from "../js-estimateMaxSpendable";
import { getTransactionStatus } from "../js-getTransactionStatus";
import { signOperation } from "../js-signOperation";
import { broadcast } from "../js-broadcast";
import type { AccountBridge, CurrencyBridge } from "@ledgerhq/types-live";

const receive = makeAccountBridgeReceive();

const updateTransaction = (t, patch) => {
  return { ...t, ...patch };
};

const preload = async () => Promise.resolve({});

const hydrate = () => {};

const currencyBridge: CurrencyBridge = {
  preload,
  hydrate,
  scanAccounts,
};

const accountBridge: AccountBridge<AlgorandTransaction> = {
  createTransaction,
  updateTransaction,
  prepareTransaction,
  getTransactionStatus,
  sync,
  receive,
  signOperation,
  broadcast,
  estimateMaxSpendable,
};

export default {
  currencyBridge,
  accountBridge,
};
