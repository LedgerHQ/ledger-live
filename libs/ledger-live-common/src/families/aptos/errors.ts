import { createCustomErrorClass } from "@ledgerhq/errors";

export const SequenceNumberTooOldError = createCustomErrorClass("SequenceNumberTooOld");

export const SequenceNumberTooNewError = createCustomErrorClass("SequenceNumberTooNew");

export const TransactionExpiredError = createCustomErrorClass("TransactionExpired");
