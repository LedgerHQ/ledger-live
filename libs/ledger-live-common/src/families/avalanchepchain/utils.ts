import type { Account } from "@ledgerhq/types-live";
import BinTools from "avalanche/dist/utils/bintools";

export const binTools = BinTools.getInstance();

export const AVAX_HRP = "avax";

export const FIGMENT_AVALANCHE_VALIDATOR_NODES = [
  "NodeID-GW7CvXwkPFDsfFHrix1SA148NHF5ZnSXs",
];

export const ONE_AVAX = 1000000000;
export const AVAX_MINIMUM_STAKE_AMOUNT = 25000000000;

export const isDefaultValidatorNode = (nodeID: string): boolean =>
  FIGMENT_AVALANCHE_VALIDATOR_NODES.includes(nodeID);

export const canDelegate = (account: Account) =>
  account.spendableBalance.gt(AVAX_MINIMUM_STAKE_AMOUNT);

export const getReadableDate = (unixTimestamp: number, locale: string) =>
  new Date(unixTimestamp * 1000).toLocaleDateString(locale, {
    year: "numeric",
    month: "short",
    day: "numeric",
    hour: "2-digit",
    minute: "2-digit",
  });

export const MINUTE = 60;
export const FIVE_MINUTES = MINUTE * 5;
export const FIFTEEN_MINUTES = MINUTE * 15;
export const HOUR = MINUTE * 60;
export const DAY = HOUR * 24;
export const YEAR = DAY * 365;
export const TWO_WEEKS = DAY * 14;
export const THREE_WEEKS = DAY * 21;
