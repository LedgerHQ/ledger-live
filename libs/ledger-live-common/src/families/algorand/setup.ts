// Goal of this file is to inject all necessary device/signer dependency to coin-modules

import { createBridges } from "@ledgerhq/coin-algorand/bridge/js";
import makeCliTools from "@ledgerhq/coin-algorand/cli-transaction";
import algorandResolver from "@ledgerhq/coin-algorand/hw-getAddress";
import { Transaction } from "@ledgerhq/coin-algorand/types";
import Algorand from "@ledgerhq/hw-app-algorand";
import Transport from "@ledgerhq/hw-transport";
import network from "@ledgerhq/live-network/network";
import type { Bridge } from "@ledgerhq/types-live";
import { CreateSigner, createResolver, executeWithSigner } from "../../bridge/setup";
import type { Resolver } from "../../hw/getAddress/types";

const createSigner: CreateSigner<Algorand> = (transport: Transport) => {
  return new Algorand(transport);
};

const bridge: Bridge<Transaction> = createBridges(executeWithSigner(createSigner), network);

const resolver: Resolver = createResolver(createSigner, algorandResolver);

const cliTools = makeCliTools();

export { bridge, cliTools, resolver };
