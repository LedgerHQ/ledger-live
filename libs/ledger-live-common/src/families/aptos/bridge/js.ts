import type { AccountBridge, CurrencyBridge } from "@ledgerhq/types-live";
import { getSerializedAddressParameters } from "@ledgerhq/coin-framework/bridge/jsHelpers";
import type { Transaction } from "../types";
import { makeAccountBridgeReceive } from "../../../bridge/jsHelpers";

import { sync, scanAccounts } from "../synchronisation";
import getTransactionStatus from "../getTransactionStatus";
import prepareTransaction from "../prepareTransaction";
import createTransaction from "../createTransaction";
import estimateMaxSpendable from "../estimateMaxSpendable";
import signOperation from "../signOperation";
import broadcast from "../broadcast";

const currencyBridge: CurrencyBridge = {
  preload: () => Promise.resolve({}),
  hydrate: () => {},
  scanAccounts,
};

const updateTransaction = (t: Transaction, patch: Partial<Transaction>): Transaction => ({
  ...t,
  ...patch,
});

const receive = makeAccountBridgeReceive();

const accountBridge: AccountBridge<Transaction> = {
  estimateMaxSpendable,
  createTransaction,
  updateTransaction,
  getTransactionStatus,
  getSerializedAddressParameters,
  prepareTransaction,
  sync,
  receive,
  signOperation,
  broadcast,
};

export default { currencyBridge, accountBridge };
