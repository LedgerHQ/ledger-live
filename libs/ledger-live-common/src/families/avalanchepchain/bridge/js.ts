import type { AccountBridge, CurrencyBridge } from "@ledgerhq/types-live";
import type { Transaction } from "../types";
import { makeAccountBridgeReceive } from "../../../bridge/jsHelpers";
import { sync, scanAccounts } from "../js-synchronisation";
import {
  createTransaction,
  prepareTransaction,
  updateTransaction,
} from "../js-transaction";
import getTransactionStatus from "../js-getTransactionStatus";
import signOperation from "../js-signOperation";
import broadcast from "../js-broadcast";
import { preload, hydrate } from "../js-preload-data";
import estimateMaxSpendable from "../js-estimateMaxSpendable";

const receive = makeAccountBridgeReceive();

const currencyBridge: CurrencyBridge = {
  preload,
  hydrate,
  scanAccounts,
};

const accountBridge: AccountBridge<Transaction> = {
  estimateMaxSpendable,
  createTransaction,
  updateTransaction,
  getTransactionStatus,
  prepareTransaction,
  sync,
  receive,
  signOperation,
  broadcast,
};
export default {
  currencyBridge,
  accountBridge,
};
