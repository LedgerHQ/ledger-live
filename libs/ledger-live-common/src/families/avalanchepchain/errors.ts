import { createCustomErrorClass } from "@ledgerhq/errors";

export const AvalancheInvalidDateTimeError = createCustomErrorClass(
  "AvalancheInvalidDateTimeError"
);
