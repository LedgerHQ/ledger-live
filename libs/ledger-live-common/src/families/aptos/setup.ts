// Goal of this file is to inject all necessary device/signer dependency to coin-modules

import {
  AptosAccount,
  TransactionStatus,
  createBridges,
  type Transaction,
} from "@ledgerhq/coin-aptos";
import Transport from "@ledgerhq/hw-transport";
import Aptos from "@ledgerhq/hw-app-aptos";
import type { Bridge } from "@ledgerhq/types-live";
import aptosResolver from "@ledgerhq/coin-aptos/signer/index";
import makeCliTools from "@ledgerhq/coin-aptos/test/cli";
import { CreateSigner, createResolver, executeWithSigner } from "../../bridge/setup";
import { Resolver } from "../../hw/getAddress/types";

const createSigner: CreateSigner<Aptos> = (transport: Transport) => {
  return new Aptos(transport);
};

const bridge: Bridge<Transaction, AptosAccount, TransactionStatus> = createBridges(
  executeWithSigner(createSigner),
);

const resolver: Resolver = createResolver(createSigner, aptosResolver);

const cliTools = makeCliTools();

export { bridge, cliTools, resolver };
