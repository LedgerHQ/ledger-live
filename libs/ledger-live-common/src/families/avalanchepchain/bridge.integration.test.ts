import "../../__tests__/test-helpers/setup";
import { testBridge } from "../../__tests__/test-helpers/bridge";
import type { DatasetTest, CurrenciesData } from "@ledgerhq/types-live";
import type { Transaction } from "./types";
import { BigNumber } from "bignumber.js";
import {
  InvalidAddressBecauseDestinationIsAlsoSource,
  AmountRequired,
  NotEnoughBalanceToDelegate,
} from "@ledgerhq/errors";
import { FIVE_MINUTES, TWO_WEEKS, AVAX_MINIMUM_STAKE_AMOUNT } from "./utils";
import { AvalancheInvalidDateTimeError } from "./errors";

const avalanchepchain: CurrenciesData<Transaction> = {
  scanAccounts: [
    {
      name: "avalanchepchain seed 1",
      apdus: `
      => 8002000015058000002c80002328800000000000000000000000
      <= 0b2d7225e42ebd50a830abd9b0bfa77555e89ebf9000
      => 800300000d038000002c8000232880000000
      <= 4104512c844e8140a7e03e6967276ef0395cc662d2b059440f7f2532ccdc9c8ad214623ccf368d78d8c06ff4e8e4436199bfc9d561304059e71b296dd0705c32906b20d31aa2c102c05840cab783582ddabafa47401c3d1d6ace40cd0ca19426722e7a9000
      `,
    },
  ],
  FIXME_ignorePreloadFields: ["validators"],
  accounts: [
    {
      FIXME_tests: [
        "balance is sum of ops",
        "invalid recipient have a recipientError",
      ],
      raw: {
        id: "js:1:avalanchepchain:avax10u6rtayzscc8ds4l5tdyjm2uazh0enhj8gguh5:",
        seedIdentifier: "",
        xpub: "xpub6Cfxnaft2T7TZ5sr3istgn2uvjTzunaZdL1rGpNU3orLUQggKYFxHSo3Csf98QwJA7BASwmGWNTwu1tc7MMRh3TvkJ6aDV55f7re6wnj42t",
        derivationMode: "",
        index: 0,
        freshAddress: "avax10u6rtayzscc8ds4l5tdyjm2uazh0enhj8gguh5",
        freshAddressPath: "44'/9000'/0'/0/0",
        freshAddresses: [
          {
            address: "avax10u6rtayzscc8ds4l5tdyjm2uazh0enhj8gguh5",
            derivationPath: "44'/9000'/0'/0/0",
          },
        ],
        name: "Avalanche P-chain 1 - Nano X",
        balance: "0",
        blockHeight: 2168241,
        currencyId: "avalanchepchain",
        unitMagnitude: 9,
        operationsCount: 0,
        operations: [],
        pendingOperations: [],
        lastSyncDate: "",
      },
      transactions: [
        {
          name: "Same as Recipient",
          transaction: (t) => ({
            ...t,
            amount: new BigNumber(100),
            recipient: "avax10u6rtayzscc8ds4l5tdyjm2uazh0enhj8gguh5",
          }),
          expectedStatus: {
            errors: {
              recipient: new InvalidAddressBecauseDestinationIsAlsoSource(),
            },
            warnings: {},
          },
        },
        {
          name: "Delegation - Not enough balance to delegate",
          transaction: (t) => ({
            ...t,
            mode: "delegate",
            amount: new BigNumber(100),
          }),
          expectedStatus: {
            errors: {
              amount: new NotEnoughBalanceToDelegate(),
            },
            warnings: {},
          },
        },
        {
          name: "Delegation - AmountRequired",
          transaction: (t) => ({
            ...t,
            mode: "delegate",
            amount: new BigNumber(0),
          }),
          expectedStatus: {
            errors: {
              amount: new AmountRequired(),
            },
            warnings: {},
          },
        },
        {
          name: "Delegation - Under time threshold",
          transaction: (t) => ({
            ...t,
            mode: "delegate",
            amount: new BigNumber(AVAX_MINIMUM_STAKE_AMOUNT),
            recipient: "GW7CvXwkPFDsfFHrix1SA148NHF5ZnSXs",
            endTime: new BigNumber(
              Math.round(new Date().getTime() / 1000) + TWO_WEEKS
            ),
          }),
          expectedStatus: {
            errors: {
              time: new AvalancheInvalidDateTimeError(),
            },
            warnings: {},
          },
        },
        {
          name: "Delegation - Beyond time threshold",
          transaction: (t) => ({
            ...t,
            mode: "delegate",
            amount: new BigNumber(AVAX_MINIMUM_STAKE_AMOUNT),
            recipient: "GW7CvXwkPFDsfFHrix1SA148NHF5ZnSXs",
            endTime: new BigNumber(
              Math.round(new Date().getTime() / 1000) + TWO_WEEKS + FIVE_MINUTES
            ),
            maxEndTime: new BigNumber(
              Math.round(new Date().getTime() / 1000) + TWO_WEEKS
            ),
          }),
          expectedStatus: {
            errors: {
              time: new AvalancheInvalidDateTimeError(),
            },
            warnings: {},
          },
        },
      ],
    },
  ],
};

const dataset: DatasetTest<Transaction> = {
  implementations: ["js"],
  currencies: {
    avalanchepchain,
  },
};

testBridge(dataset);
