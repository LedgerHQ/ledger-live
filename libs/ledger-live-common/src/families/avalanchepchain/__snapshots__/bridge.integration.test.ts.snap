// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`avalanchepchain currency bridge scanAccounts avalanchepchain seed 1 1`] = `
Array [
  Object {
    "avalanchePChainResources": Object {
      "chainCode": "d31aa2c102c05840cab783582ddabafa47401c3d1d6ace40cd0ca19426722e7a",
      "delegations": Array [],
      "publicKey": "03512c844e8140a7e03e6967276ef0395cc662d2b059440f7f2532ccdc9c8ad214",
      "stakedBalance": "0",
    },
    "balance": "75564408151",
    "currencyId": "avalanchepchain",
    "derivationMode": "avalanchepchain",
    "freshAddress": "avax1pvkhyf0y9674p2ps40vmp0a8w427384lpr8zan",
    "freshAddressPath": "44'/9000'/0'/0/0",
    "freshAddresses": Array [
      Object {
        "address": "avax1pvkhyf0y9674p2ps40vmp0a8w427384lpr8zan",
        "derivationPath": "44'/9000'/0'/0/0",
      },
    ],
    "id": "js:2:avalanchepchain:avax1pvkhyf0y9674p2ps40vmp0a8w427384lpr8zan:avalanchepchain",
    "index": 0,
    "name": "Avalanche (P-Chain) 1",
    "nfts": undefined,
    "operationsCount": 11,
    "pendingOperations": Array [],
    "seedIdentifier": "03512c844e8140a7e03e6967276ef0395cc662d2b059440f7f2532ccdc9c8ad214",
    "spendableBalance": "75564408151",
    "starred": false,
    "swapHistory": Array [],
    "syncHash": undefined,
    "unitMagnitude": 9,
    "used": true,
  },
]
`;

exports[`avalanchepchain currency bridge scanAccounts avalanchepchain seed 1 2`] = `
Array [
  Array [
    Object {
      "accountId": "js:2:avalanchepchain:avax1pvkhyf0y9674p2ps40vmp0a8w427384lpr8zan:avalanchepchain",
      "blockHash": "E5MxxvcEAZBA9SKTcDcacBp8yL8AYUu65SPozZL94rqDTTmwn",
      "blockHeight": 1921983,
      "contract": undefined,
      "extra": Object {},
      "fee": "1000000",
      "hash": "24etLv7LZ7wLpoWQnqD36aodZQ1LKZbTLYC6oaoWtJythADzZy",
      "id": "js:2:avalanchepchain:avax1pvkhyf0y9674p2ps40vmp0a8w427384lpr8zan:avalanchepchain-24etLv7LZ7wLpoWQnqD36aodZQ1LKZbTLYC6oaoWtJythADzZy-IN",
      "operator": undefined,
      "recipients": Array [
        "avax1kh9r03972vqxlglsarxrc9f9mxe2zcw5prfhel",
      ],
      "senders": Array [],
      "standard": undefined,
      "tokenId": undefined,
      "type": "IN",
      "value": "1000000000",
    },
    Object {
      "accountId": "js:2:avalanchepchain:avax1pvkhyf0y9674p2ps40vmp0a8w427384lpr8zan:avalanchepchain",
      "blockHash": "dcb81QQ2jx5SAr5BsWmgepEiWSy8RfSwupredqzZZke5w76Mu",
      "blockHeight": 1932215,
      "contract": undefined,
      "extra": Object {
        "stakeValue": "25000000000",
      },
      "fee": "0",
      "hash": "2a9qJQvitJd1Y5RiqLoMonxVEj76PzvusXgHD1GztYjJUXtJmL",
      "id": "js:2:avalanchepchain:avax1pvkhyf0y9674p2ps40vmp0a8w427384lpr8zan:avalanchepchain-2a9qJQvitJd1Y5RiqLoMonxVEj76PzvusXgHD1GztYjJUXtJmL-DELEGATE",
      "operator": undefined,
      "recipients": Array [],
      "senders": Array [],
      "standard": undefined,
      "tokenId": undefined,
      "type": "DELEGATE",
      "value": "0",
    },
    Object {
      "accountId": "js:2:avalanchepchain:avax1pvkhyf0y9674p2ps40vmp0a8w427384lpr8zan:avalanchepchain",
      "blockHash": "LpMqaFyUa9zk9ZfVv1fapQQN1ghvGExxZCkd4BDK2CtrjVKeA",
      "blockHeight": 1610802,
      "contract": undefined,
      "extra": Object {},
      "fee": "1000000",
      "hash": "2Gkti7XsfVzrM4okeBMTokAVwXyWRgiQ6Wmj83E3q1tRAMkdJi",
      "id": "js:2:avalanchepchain:avax1pvkhyf0y9674p2ps40vmp0a8w427384lpr8zan:avalanchepchain-2Gkti7XsfVzrM4okeBMTokAVwXyWRgiQ6Wmj83E3q1tRAMkdJi-IN",
      "operator": undefined,
      "recipients": Array [
        "avax1pvkhyf0y9674p2ps40vmp0a8w427384lpr8zan",
      ],
      "senders": Array [],
      "standard": undefined,
      "tokenId": undefined,
      "type": "IN",
      "value": "100000000",
    },
    Object {
      "accountId": "js:2:avalanchepchain:avax1pvkhyf0y9674p2ps40vmp0a8w427384lpr8zan:avalanchepchain",
      "blockHash": "Ym9dPKgTK7w4qskJQM5Bm7pPmyWAw691XBPdx4CUTatpDP7af",
      "blockHeight": 1922043,
      "contract": undefined,
      "extra": Object {},
      "fee": "1000000",
      "hash": "2kDCa96WgnT9T4MtXABJZWzU4RnafDnEEN6o4RNV44L8zy9CYJ",
      "id": "js:2:avalanchepchain:avax1pvkhyf0y9674p2ps40vmp0a8w427384lpr8zan:avalanchepchain-2kDCa96WgnT9T4MtXABJZWzU4RnafDnEEN6o4RNV44L8zy9CYJ-IN",
      "operator": undefined,
      "recipients": Array [
        "avax10ahyx8xnnapxll8ny0ttvs3uycnm6ks8p3ker4",
      ],
      "senders": Array [],
      "standard": undefined,
      "tokenId": undefined,
      "type": "IN",
      "value": "24000000000",
    },
    Object {
      "accountId": "js:2:avalanchepchain:avax1pvkhyf0y9674p2ps40vmp0a8w427384lpr8zan:avalanchepchain",
      "blockHash": "yNpemx5Dv3WnS6Xnhyh8UT1CJYPRLtT8NSAZKZE1ydBog7bRx",
      "blockHeight": 1632371,
      "contract": undefined,
      "extra": Object {},
      "fee": "1000000",
      "hash": "2Lw6Liyo4DufE85fYhTHV8EdE3PvtptT1gCJAUfxTff2P3CdDf",
      "id": "js:2:avalanchepchain:avax1pvkhyf0y9674p2ps40vmp0a8w427384lpr8zan:avalanchepchain-2Lw6Liyo4DufE85fYhTHV8EdE3PvtptT1gCJAUfxTff2P3CdDf-IN",
      "operator": undefined,
      "recipients": Array [
        "avax1p240en83g86whf929pfjq4tp4ynkgzgv5xw6z3",
      ],
      "senders": Array [],
      "standard": undefined,
      "tokenId": undefined,
      "type": "IN",
      "value": "300000000",
    },
    Object {
      "accountId": "js:2:avalanchepchain:avax1pvkhyf0y9674p2ps40vmp0a8w427384lpr8zan:avalanchepchain",
      "blockHash": "2URZyYLE2cb4WMhWs6xCN1ei9x3X16VuNjrS1TPc2R5D6u7hQ7",
      "blockHeight": 1929061,
      "contract": undefined,
      "extra": Object {},
      "fee": "1000000",
      "hash": "2uGfC7MdRC394CRKtWjAqp6dW2yh32xK3qJ3LD3EKjRFtSBoxt",
      "id": "js:2:avalanchepchain:avax1pvkhyf0y9674p2ps40vmp0a8w427384lpr8zan:avalanchepchain-2uGfC7MdRC394CRKtWjAqp6dW2yh32xK3qJ3LD3EKjRFtSBoxt-IN",
      "operator": undefined,
      "recipients": Array [
        "avax1nne34ukaqsgqxmjgsuup8ymt08lulj5546g7c6",
      ],
      "senders": Array [],
      "standard": undefined,
      "tokenId": undefined,
      "type": "IN",
      "value": "25000000000",
    },
    Object {
      "accountId": "js:2:avalanchepchain:avax1pvkhyf0y9674p2ps40vmp0a8w427384lpr8zan:avalanchepchain",
      "blockHash": "nEVxJJ7kw1FEGox5YsRJQ9PqCeSvfHSeprRTZjvc4BhKEM9vv",
      "blockHeight": 1611438,
      "contract": undefined,
      "extra": Object {},
      "fee": "1000000",
      "hash": "BSVFaYseLC9QUuUTtYGxCh8tf7DE7zqHpwZPBBH6CSwiutUv6",
      "id": "js:2:avalanchepchain:avax1pvkhyf0y9674p2ps40vmp0a8w427384lpr8zan:avalanchepchain-BSVFaYseLC9QUuUTtYGxCh8tf7DE7zqHpwZPBBH6CSwiutUv6-OUT",
      "operator": undefined,
      "recipients": Array [
        "avax1rlu93pnalclt7h0dte3evua97lv4fp4qzax5wq",
      ],
      "senders": Array [
        "avax1pvkhyf0y9674p2ps40vmp0a8w427384lpr8zan",
      ],
      "standard": undefined,
      "tokenId": undefined,
      "type": "OUT",
      "value": "12000000",
    },
    Object {
      "accountId": "js:2:avalanchepchain:avax1pvkhyf0y9674p2ps40vmp0a8w427384lpr8zan:avalanchepchain",
      "blockHash": "hXnSceGbJajBQeKsFypokezZroD87LRF75yCsGEBoX2p4CQ5Y",
      "blockHeight": 1924119,
      "contract": undefined,
      "extra": Object {
        "stakeValue": "25000000000",
      },
      "fee": "0",
      "hash": "E82EwpekLxZ9DnKQquCEVR9KR6G3tF1ut3xhUjLcKjtVZQ4GE",
      "id": "js:2:avalanchepchain:avax1pvkhyf0y9674p2ps40vmp0a8w427384lpr8zan:avalanchepchain-E82EwpekLxZ9DnKQquCEVR9KR6G3tF1ut3xhUjLcKjtVZQ4GE-DELEGATE",
      "operator": undefined,
      "recipients": Array [],
      "senders": Array [],
      "standard": undefined,
      "tokenId": undefined,
      "type": "DELEGATE",
      "value": "0",
    },
    Object {
      "accountId": "js:2:avalanchepchain:avax1pvkhyf0y9674p2ps40vmp0a8w427384lpr8zan:avalanchepchain",
      "blockHash": "2bc3KP8cWuTbrWarJbuMAoqGQwySkzouSMFZ9SoKg7s9RSRMCg",
      "blockHeight": 1990187,
      "contract": undefined,
      "extra": Object {
        "stakeValue": "25381000000",
      },
      "fee": "0",
      "hash": "fr3Eq8FeKN1YGVAKr1DEc3hAbCSBqrVXrG2Pwf8nuCXySDUrw",
      "id": "js:2:avalanchepchain:avax1pvkhyf0y9674p2ps40vmp0a8w427384lpr8zan:avalanchepchain-fr3Eq8FeKN1YGVAKr1DEc3hAbCSBqrVXrG2Pwf8nuCXySDUrw-DELEGATE",
      "operator": undefined,
      "recipients": Array [],
      "senders": Array [],
      "standard": undefined,
      "tokenId": undefined,
      "type": "DELEGATE",
      "value": "0",
    },
    Object {
      "accountId": "js:2:avalanchepchain:avax1pvkhyf0y9674p2ps40vmp0a8w427384lpr8zan:avalanchepchain",
      "blockHash": "vo5vFaXhCiivQAhKb36du7uE3mdadDu8ij4gurstStEPMmB3g",
      "blockHeight": 1937708,
      "contract": undefined,
      "extra": Object {},
      "fee": "1000000",
      "hash": "PEe7C3JaqeWthivMHEnKMkYNedZP8S2qgVySaFQRajHPCu1hQ",
      "id": "js:2:avalanchepchain:avax1pvkhyf0y9674p2ps40vmp0a8w427384lpr8zan:avalanchepchain-PEe7C3JaqeWthivMHEnKMkYNedZP8S2qgVySaFQRajHPCu1hQ-IN",
      "operator": undefined,
      "recipients": Array [
        "avax1s8y9hyyyrtjunt3a477wn5kcpk3tq2aln7yk8c",
      ],
      "senders": Array [],
      "standard": undefined,
      "tokenId": undefined,
      "type": "IN",
      "value": "25000000000",
    },
    Object {
      "accountId": "js:2:avalanchepchain:avax1pvkhyf0y9674p2ps40vmp0a8w427384lpr8zan:avalanchepchain",
      "blockHash": "NYfn1W5KWfZZ484MV6sRMqH9TvXTxS1FCQsuYjVnFNSEvBGgh",
      "blockHeight": 1683042,
      "contract": undefined,
      "extra": Object {},
      "fee": "1000000",
      "hash": "UsXVfL8uj3Kv4t6h7fvy4nPNXkf7q4HgZLkDoTnsMzCF5DcKo",
      "id": "js:2:avalanchepchain:avax1pvkhyf0y9674p2ps40vmp0a8w427384lpr8zan:avalanchepchain-UsXVfL8uj3Kv4t6h7fvy4nPNXkf7q4HgZLkDoTnsMzCF5DcKo-OUT",
      "operator": undefined,
      "recipients": Array [
        "avax1nhx84207m2t85pyud5lpyucaja9zg0amt62zfl",
      ],
      "senders": Array [
        "avax1p240en83g86whf929pfjq4tp4ynkgzgv5xw6z3",
      ],
      "standard": undefined,
      "tokenId": undefined,
      "type": "OUT",
      "value": "7000000",
    },
  ],
]
`;
